pilot/test/cloudfoundry/integration/copilot_test.go

```diff
@@ -0,0 +1,323 @@
+
+	t.Log("checking if pilot received routes from copilot")
+	g.Eventually(func() (string, error) {
+		return curlPilot(fmt.Sprintf("http://127.0.0.1:%d/v1/registration", pilotPort))
+	}).Should(gomega.ContainSubstring("abc.com"))
+
+	g.Eventually(func() (string, error) {
+		return curlPilot(fmt.Sprintf("http://127.0.0.1:%d/v1/listeners/x/router~x~x~x", pilotPort))
+	}).Should(gomega.ContainSubstring("http_connection_manager"))
+
+	t.Log("docker run envoy...")
+	proxySession, err := testState.runEnvoyDocker(pilotPort, localIP, envoyImageName)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	testState.addCleanupTask(func() {
+		proxySession.Terminate()
+		g.Eventually(proxySession, "5s").Should(gexec.Exit())
+	})
+
+	g.Consistently(proxySession).ShouldNot(gexec.Exit())
+
+	t.Log("curling the app with expected host header")
+	g.Eventually(func() error {
+		respData, err := curlApp(fmt.Sprintf("http://%s:%d", "127.0.0.1", 8080), "abc.com")
+		if err != nil {
+			return err
+		}
+		if respData != "hello" {
+			return fmt.Errorf("unexpected response data: %s", respData)
+		}
+		return nil
+	}, "5s", "100ms").Should(gomega.Succeed())
+}
+
+type testState struct {
+	// generated credentials for server and client
+	creds testhelpers.MTLSCredentials
+
+	// path on disk to store the config.yaml
+	configFilePath string
+
+	config *cloudfoundry.Config
+
+	cleanupTasks []func()
+	mutex        sync.Mutex
+}
+
+func (testState *testState) addCleanupTask(task func()) {
+	testState.mutex.Lock()
+	defer testState.mutex.Unlock()
+	testState.cleanupTasks = append(testState.cleanupTasks, task)
+}
+
+func newTestState(mockServerAddress string) *testState {
+	creds := testhelpers.GenerateMTLS()
+	clientTLSFiles := creds.CreateClientTLSFiles()
+	return &testState{
+		creds:          creds,
+		configFilePath: filepath.Join(creds.TempDir, "config.yaml"),
+		config: &cloudfoundry.Config{
+			Copilot: cloudfoundry.CopilotConfig{
+				ServerCACertPath: clientTLSFiles.ServerCA,
+				ClientCertPath:   clientTLSFiles.ClientCert,
+				ClientKeyPath:    clientTLSFiles.ClientKey,
+				Address:          mockServerAddress,
+				PollInterval:     10 * time.Second,
+			},
+		},
+	}
+}
+
+func (testState *testState) runEnvoyDocker(pilotPort int, localIP, envoyImageName string) (*gexec.Session, error) {
+	tempDir, err := ioutil.TempDir("", "istio-cloudfoundry-test")
+	if err != nil {
+		return nil, err
+	}
+	tempDir, err = filepath.EvalSymlinks(tempDir)
+	if err != nil {
+		return nil, err
+	}
+	testState.addCleanupTask(func() { os.RemoveAll(tempDir) })
+
+	err = writeEnvoyConfig(filepath.Join(tempDir, "envoy.json"), pilotPort, localIP)
+	if err != nil {
+		return nil, err
+	}
+
+	cmd := exec.Command("docker", "run",
+		"-v", fmt.Sprintf("%s:%s", tempDir, "/envoy-config"),
+		"-p", "8080:8080",
+		"--entrypoint", "/usr/local/bin/envoy",
+		envoyImageName,
+		"-c", "/envoy-config/envoy.json",
+		"--service-cluster", "x",
+		"--service-node", "router~x~x~x",
```

Comment:
I think we have code in pilot/proxy/envoy/*_test, that uses an envoy binary to test, instead of a docker container. If you reuse that code, you won't have to worry about which docker container or binary to use.

Reply:
Plus, its more lightweight